docs/adrs/003-Retry_endpoint_design.md

```diff
@@ -0,0 +1,161 @@
+
+#### Q: How do we find the most recent retry of a job?
+A: The client using the ee2 API would have to figure it out using the `retry_parent` and job creation date fields. (Unless we added other fields to help with this)
+
+#### Q: How do we ensure that the app version is correctly run each time when submitting from the narrative?
+A: We would need to change the narrative to submit the git commit hash instead of a version tag
+
+#### Q: How do we handle DB consistency during retry failure? 
+Looks like the options are
+* implement db integrity checks and two-phase commits for making the relationships between a job, its `retry_parent`, and the batch container
+* accept that the db info may be incomplete and write workarounds into the clients
+* (upgrade to Mongo 4.4 for better transaction support)
+A: We have decided to use a `retry_toggle` in order to mark that the entire transaction has occurred for a retry job, and to set up a monitor to fix the jobs that didn't finish the retry lifecycle.
+
+##### Q: Do we want to support ResourceRequirements
+A: Probably not in the short term
+
+
+#### Q: how to prevent incorrect parent-child relationships being created -- should the client be allowed to specify a parent ID? Is it currently possible to add a new child to a parent job if the child is a new job, rather than an existing job ID / set of params that is being rerun?
+A: Not necessarily relevant to this endpoint, more of a `run_job_batch` endpoint question. Currently the `retry_parent` and `parent_job_id` are looked up from the ee2 record on retry, and not specified in this endpoint.
+
+#### Shorter Q and A
+
+    Should we track a retry count? (Done)
+    Should users see this retry count? (Unknown TBD)
+    Are retried jobs saved in some sort of data structure linking them, possibly indirectly, to the parent job or are they orphaned? (Yes, retry_parent)
+    If the former, is the retry relationship linear or a tree? E.g. what happens if there are two simultaneous calls to retry a job? (Tree, simultaneous jobs run)
+    Should it be at least theoretically possible to see the list of retried jobs in order? (It is possible by sorting on creation date)
+    Should there be a maximum retry count? Or a warning that more retries are not likely to help?  (Unknown TBD)
+    Can a job in states other than failed or canceled be retried? Or should the user be required to cancel a job before it can be retried? (Job must be in Error/Cancel state)
+
+
+# Work estimation
+Priority descending
+* Address data inconsistency via retry_count, retry_ids and retry_toggle
+> Estimate 3-4 days
+> https://kbase-jira.atlassian.net/browse/DATAUP-461
+
+* Preflight checks
+> Estimate 3-4 days
+> https://kbase-jira.atlassian.net/browse/DATAUP-528
+
+
+
+
+> Requires retry to be able to force the same app `git_commit versions` and `JobRequirements` from the db records
+https://kbase-jira.atlassian.net/browse/DATAUP-461
```

Comment:
Do these two lines go with Preflight checks above or Hooup retries below?

Reply:
Reformatted

Reply:
Ok, so I thought we weren't going to to this, at least not as part of the MVP. 

Maybe the issue is that the work estimates imply that we're going to do the work - how about putting a disclaimer at the top of the work estimates section that that's not the case?

Reply:
ok done